main/src/ca/uwaterloo/flix/util/collection/Chain.scala

```diff
@@ -47,34 +47,97 @@ sealed trait Chain[+A] extends Iterable[A] {
     * The empty chain.
     */
   override val empty: Chain[A] = Chain.Empty
+
+  /**
+    * Returns the amount of elements in the chain.
+    */
+  def length: Int = this match {
+    case Chain.Empty => 0
+    case Chain.Link(l, r) => l.length + r.length
+    case Chain.Many(cs) => cs.map(_.length).sum
+    case Chain.Proxy(xs) => xs.length
+  }
+
+  override def hashCode(): Int = this.toList.hashCode()
+
+  override def equals(obj: Any): Boolean = obj match {
+    case that: Chain[_] => this.toList == that.toList
+    case _ => false
+  }
 }
 
 object Chain {
 
   /**
     * The empty chain.
     */
-  private case object Empty extends Chain[Nothing]
+  private case object Empty extends Chain[Nothing] {
+
+    /**
+      * Returns `this` as a [[List]].
+      *
+      * NB.: This method must be overridden in each case class,
+      * otherwise it results in a [[StackOverflowError]].
+      */
+    override def toList: List[Nothing] = Nil
```

Comment:
Why would a match cause an overflow?

Reply:
No clue but there seems to be some kind of circular pattern going on in `Iterable` when not overriding `toList` (also when overriding it in the trait instead of each case class) that causes it